### PR TITLE
feat(tools): add completion verification hook

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

- Adds a configurable pre-done guard to the `complete` tool: before closing the session, run a shell command to verify the work actually passes
- If the command exits non-zero, the agent receives a user-role repair prompt with the failure output treated as untrusted diagnostic data and gets another turn to fix the issue
- After `GPTME_VERIFY_COMPLETION_MAX_RETRIES` failed attempts (default: 3) the hook closes anyway to prevent infinite loops
- 39 new tests covering success/failure/retry/timeout paths, workspace-script discovery, confirmation, process cleanup, TOCTOU safety, and prompt-injection boundaries

## Configuration

```bash
# Via environment variable
export GPTME_VERIFY_COMPLETION="pytest tests/ -x"
gptme "fix the failing tests"

# Via workspace script (auto-detected, no env var needed)
cat .gptme/verify-completion.sh
#!/bin/sh
pytest tests/ -x && mypy src/
```

Additional knobs:
- `GPTME_VERIFY_COMPLETION_MAX_RETRIES` (default: 3)
- `GPTME_VERIFY_COMPLETION_TIMEOUT` (default: 60s)

## Motivation

Cross-harness demand signal from rival harness repo mining: Cline users repeatedly report premature `attempt_completion` calls (Cline#11546); OpenHands users want audit evidence for every completed action (OpenHands#15377). This adds the same guard to gptme with zero overhead when not configured.

## Test plan

- [x] `pytest tests/test_tools_complete.py::TestCompleteHookVerification` — 39 tests, all passing
- [x] Full `pytest tests/test_tools_complete.py` suite — 102 tests, all passing
- [x] Pre-commit hooks (`prek`) clean — ruff, mypy, format all pass
- [x] CI green

### CODE REVIEW

File: gptme/tools/complete.py [hunk 1]
```diff
@@ -92,6 +141,62 @@ def complete_hook(
...
+                max_retries = _env_int(
+                    "GPTME_VERIFY_COMPLETION_MAX_RETRIES", _DEFAULT_MAX_RETRIES
+                )
+                # Count prior complete-call attempts via durable "Task complete"
+                # system messages — each failed verification leaves one in the log.
+                prior_attempts = sum(
+                    1
+                    for m in messages
+                    if m.role == "system" and _TASK_COMPLETE_MSG in (m.content or "")
+                )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in ef2a51316: switched the retry counter from _TASK_COMPLETE_MSG (which already includes the current attempt by the time the hook runs) to counting prior _VERIFY_FAILED_MARKER messages, which are only ever written by an earlier run of this hook. Verified with a new test that simulates 3 consecutive failed attempts and confirms the 4th call is the one that gives up.

File: gptme/tools/complete.py [hunk 2]
```diff
@@ -16,6 +18,45 @@
...
+    1. ``GPTME_VERIFY_COMPLETION`` environment variable
+    2. Executable file at ``<workspace>/.gptme/verify-completion.sh``
+    """
+    cmd = os.environ.get("GPTME_VERIFY_COMPLETION")
+    if cmd:
+        return cmd
+    if workspace is not None:
+        script = workspace / ".gptme" / "verify-completion.sh"
+        if script.is_file() and os.access(script, os.X_OK):
+            return str(script)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in ef2a51316: the workspace-controlled `.gptme/verify-completion.sh` path now goes through the same `confirm()` gate ordinary shell tool calls use (gptme.hooks.confirm), so it's subject to CLI/server confirmation hooks instead of running unconditionally. The operator-configured `GPTME_VERIFY_COMPLETION` env var path is unaffected (that's explicit operator config, not repo-controlled content).

File: gptme/tools/complete.py [hunk 3]
```diff
@@ -16,6 +18,45 @@
...
+    timeout = _env_int("GPTME_VERIFY_COMPLETION_TIMEOUT", _DEFAULT_VERIFY_TIMEOUT)
+    return subprocess.run(
+        cmd,
+        shell=True,
+        capture_output=True,
+        text=True,
+        cwd=workspace,
+        timeout=timeout,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in ef2a51316: the verify command now runs via Popen with start_new_session=True (its own process group), and on timeout we os.killpg() the whole group instead of just terminating the immediate shell, so spawned descendants (e.g. test workers) don't keep running.

**Comment by TimeToBuildBob** [comment]
Automated placeholder: work was recorded for this monitoring item but the worker finished before posting the thread update. See evidence below for what changed — the delivery guard is NOT closing the loop, just preventing silent orphaning.

Evidence: commit: 76da56a9cf docs(journal): gptme#3358 — fix 3 Greptile P1 findings on completion verification hook

Session: `324f6f55-c75e-59ed-bf4f-e051dac82ddf`

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
All CI checks now passing ✅ (Greptile: pass, 0 unresolved threads). Ready for human review — cannot self-merge because `gptme/tools/complete.py` is outside the allowed self-merge categories.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/complete.py [hunk 4]
```diff
@@ -92,6 +169,92 @@ def complete_hook(
...
+                        workspace=workspace,
+                    )
+                    if _confirm_result.action != ConfirmAction.CONFIRM:
+                        logger.info(
+                            "Completion verification script declined by confirmation gate: %s",
+                            verify_cmd,
+                        )
+                        raise SessionCompleteException(
+                            "Session completed via complete tool"
+                        )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in 0bba40bd4: the EDIT branch now uses `_confirm_result.edited_content` as the verify command instead of closing the session, matching how ordinary shell confirmation handles edits. Only SKIP and other non-EDIT/non-CONFIRM actions close the session. Added a test (`test_workspace_script_edited_by_confirmation_runs_edited_cmd`) that verifies the original script does NOT run and the edited command DOES run.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by ErikBjare** [comment]
@TimeToBuildBob remaining issues

**Comment by TimeToBuildBob** [comment]
Fixed in a86ccb3. The two remaining issues from the Greptile summary were:

**1. Windows timeout leaks descendants** — changed `proc.terminate()` to `taskkill /F /T /PID` which kills the entire process tree, matching the POSIX `os.killpg` behaviour.

**2. Empty EDIT content falls through to wrong branch** — added an explicit `elif action == EDIT` branch for the empty-content case that logs "skipped (empty command after edit)" instead of silently hitting the "declined" elif with a misleading log message.

Also added 3 new tests covering the timeout failure message, the Windows taskkill call (mocked), and the empty-EDIT close path. All 22 verification tests pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/complete.py [hunk 5]
```diff
@@ -92,6 +175,108 @@ def complete_hook(
...
+                # the current attempt's own message, since execute_complete has
+                # already appended it by the time this hook runs — undercounting
+                # the allowed retries by one), _VERIFY_FAILED_MARKER is only ever
+                # written by a previous run of this hook, so it never includes
+                # the current attempt.
+                prior_attempts = sum(
+                    1
+                    for m in messages
+                    if m.role == "system" and _VERIFY_FAILED_MARKER in (m.content or "")
+                )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in 3a4397ec0: the retry counter now tracks `_TASK_COMPLETE_MSG` occurrences (appended by `execute_complete` to the persistent conversation log on every `complete` call) and subtracts 1 for the current attempt's own marker.

`_VERIFY_FAILED_MARKER` was never suitable here — GENERATION_PRE hook messages are only added to a generation-time copy of the message list, not to the persistent log. Every subsequent `complete` call saw `prior_attempts = 0`, allowing the verifier to retry indefinitely.

Updated tests to match: `_COMPLETE_MSG` now includes the `_TASK_COMPLETE_MSG` that `execute_complete` appends before GENERATION_PRE fires, and the integration test appends `_system(_TASK_COMPLETE_MSG)` (not the failure message) to simulate each subsequent complete call.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed in f8948f59. The three new Greptile P1 findings from the latest review:

**1. Workspace scripts bypass denylisting** — instead of checking `is_denylisted(script_path)` (the path itself has no dangerous patterns), now reads the script file contents and passes those to `is_denylisted()`. A workspace `.gptme/verify-completion.sh` containing `rm -rf /` is blocked before Popen.

**2. Failed cleanup defeats timeout** — changed `proc.wait()` after the taskkill/killpg cleanup to `proc.wait(timeout=5)`; if that times out (cleanup failed, process still alive), calls `proc.kill()` + `proc.wait(timeout=2)` before re-raising. No longer blocks indefinitely if taskkill is unavailable.

**3. Historical completions consume retries** — replaced the all-messages `_TASK_COMPLETE_MSG` sum with an episode-scoped backward walk that stops at any non-complete assistant message or real user message. A `_TASK_COMPLETE_MSG` from a prior (resumed) session no longer consumes the current retry allowance.

Added tests for all three cases.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed in fcfbd67. The three P1 findings from the latest Greptile review:

**1. Repair turns reset retries** — the episode walk previously broke on any non-`complete` assistant turn, resetting `prior_attempts` to 0 after every repair attempt. Changed to `continue` through all assistant turns; the episode boundary is now exclusively a real user message (not auto-replies). Regression test added.

**2. Edited commands bypass denylisting** — when confirmation editing replaced the workspace script path with a shell command, `Path(verify_cmd).read_text()` raised `OSError` and `_script_content` stayed `None`, skipping `is_denylisted()`. Now falls back to checking `verify_cmd` itself on `OSError`. Regression test added.

**3. Context exit restores unbounded wait** — `with Popen(...) as proc:` calls `proc.wait()` without timeout in `__exit__`, so a Windows process that survived `proc.kill()` could hang the session indefinitely. Replaced with explicit `try/finally` that closes pipes and calls `proc.wait(timeout=1)` under `contextlib.suppress` — always bounded.

99 tests, all passing.

File: gptme/tools/complete.py [hunk 6]
```diff
@@ -92,6 +196,154 @@ def complete_hook(
...
+
+                if is_workspace_script:
+                    # The workspace-script path calls Popen directly, bypassing
+                    # execute_shell()'s is_denylisted() check. Read the script
+                    # contents and apply the same check so that destructive
+                    # commands inside repo-controlled scripts are blocked even
+                    # in autonomous / auto-confirm modes.
+                    _script_content: str | None = None
+                    try:
+                        _script_content = Path(verify_cmd).read_text()
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in b00b7beca. The hook now snapshots the workspace script before confirmation, validates that exact content, and executes the same content through the shell rather than reopening the repository-controlled path. Added a regression test that replaces the script during confirmation and verifies only the approved snapshot runs.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/complete.py [hunk 7]
```diff
@@ -92,6 +196,158 @@ def complete_hook(
...
+                        raise SessionCompleteException(
+                            "Session completed via complete tool"
+                        )
+
+                if is_workspace_script:
+                    # This path bypasses execute_shell()'s denylist check. Validate
+                    # and execute the pre-confirmation snapshot rather than reopening
+                    # a repository-controlled path that may since have changed.
+                    assert script_content is not None
+                    verify_cmd = script_content
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in a098e8bca. The approved workspace script is now written to an atomically created private executable snapshot and launched directly with `shell=False`, so the kernel honors its shebang and `$0` / `BASH_SOURCE` identify an actual script file. The hook still runs the exact content that was validated, never reopens the repository-controlled path, and deletes the snapshot after bounded process cleanup. Added a regression test using Bash-only syntax plus `$0` / `BASH_SOURCE` checks; all 101 completion-tool tests and scoped pre-commit checks pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/complete.py [hunk 8]
```diff
@@ -92,6 +231,166 @@ def complete_hook(
...
+                            verify_cmd,
+                        )
+                        yield Message(
+                            "system",
+                            f"{_VERIFY_FAILED_MARKER}: exit code {result.returncode}.\n"
+                            f"Command: `{verify_cmd}`\n"
+                            f"Output:\n{output}\n\n"
+                            f"Please fix the issue and call complete again.",
+                            quiet=False,
+                        )
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/complete.py]: Fixed in 8da8af870. Failed verifier output is now returned as a user-role repair prompt rather than a system-role instruction. The command and stdout/stderr are explicitly labelled as untrusted repository-controlled diagnostic data, XML-escaped, and enclosed in a delimiter that verifier output cannot close. Added regressions for the role boundary and an attempted `</verifier-output><system>…` escape; all 102 completion-tool tests and scoped pre-commit checks pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Addressed the remaining score-driving security issue in 8da8af870: verifier stdout/stderr no longer enters the system channel. It is now a user-role repair prompt, explicitly marked as untrusted repository-controlled diagnostic data, XML-escaped, and delimiter-contained. The new regression includes an attempted delimiter/system-tag escape; all 102 completion-tool tests and scoped pre-commit checks pass.

The inline finding is fixed and resolved. Fresh CI is running, and I triggered exactly one guarded Greptile re-review for the new head. No further bot-review loop after this cycle; if it stays at 3/5 despite no stated findings, this is ready for your judgment rather than another trigger.

**Comment by TimeToBuildBob** [comment]
All inline review threads are resolved and CI is green on the current head (`8da8af870`). Could you point me to the specific remaining concerns? I want to make sure I'm addressing what you have in mind rather than guessing at the 3/5 Greptile score.

Current state: lint ✅ typecheck ✅ tests ✅ build ✅ Greptile check ✅ — no outstanding threads.

**Comment by TimeToBuildBob** [comment]
The two remaining Greptile P1 findings (found after your comment) have been addressed:

- **Snapshot loses script semantics** (20:45 UTC) — fixed in `a098e8bca`: verifier now runs from an atomically-created private executable snapshot, preserving shebang + `$0`/`BASH_SOURCE` semantics, executing the validated content only.
- **Verifier output gains system authority** (21:46 UTC) — fixed in `8da8af870`: failed verifier stdout/stderr is now a user-role repair prompt (XML-escaped, delimiter-contained, labelled untrusted) rather than a system-role instruction.

Current state on `8da8af870`: lint ✅ typecheck ✅ tests ✅ build ✅ Greptile Review ✅ — 102 tests, all passing. Ready for merge when you're happy with the design.

**Comment by TimeToBuildBob** [comment]
Cleaned up one unrelated artifact before handoff: commit `083ad4072` drops the no-op 356-line `poetry.lock` rewrite produced by an older Poetry version. The PR is now scoped to `gptme/tools/complete.py` and `tests/test_tools_complete.py` only (905 additions, 2 deletions). Re-ran the focused suite after cleanup: 102 passed. Fresh CI is running on the new head.

**Comment by TimeToBuildBob** [comment]
All issues addressed. Here's the full fix list since your "remaining issues" comment:

**Fixes since ~19:23 UTC:**
1. **Windows timeout leaks descendants** (`a86ccb3`) — switched `proc.terminate()` to `taskkill /F /T /PID` to kill the entire process tree (matching POSIX `os.killpg`)
2. **Workspace scripts bypass denylisting** (`f8948f59`) — verifier now reads script contents and passes them to `is_denylisted()`, so a `.gptme/verify-completion.sh` containing `rm -rf /` is blocked before `Popen`
3. **Repair turns reset retries** (`fcfbd67`) — episode walk now `continue`s through all assistant turns; the episode boundary is exclusively a real user message
4. **Snapshot loses script semantics** (`a098e8bca`) — verifier runs from a private executable snapshot, preserving shebang and `$0`/`BASH_SOURCE`
5. **Verifier output gains system authority** (`8da8af870`) — failed verifier stdout/stderr is now a user-role repair prompt (XML-escaped, delimiter-contained, labelled untrusted)
6. **Unrelated poetry.lock artifact removed** (`083ad4072`) — PR is now scoped to `gptme/tools/complete.py` + `tests/test_tools_complete.py` only

All 14 Greptile inline threads resolved. CI is running on `083ad4072`. Ready for your review when CI finishes.

**Comment by TimeToBuildBob** [comment]
CI is now fully green on `083ad4072` — all 13 checks pass. Ready for merge.
